Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp

```diff
@@ -586,16 +586,19 @@ void KafkaEventStreamDecoder::initLocalCaches(
         spDetMsg->spectrum()->data(), spDetMsg->detector_id()->data(), nudet);
   }
 
-  // Load the instrument if possible but continue if we can't
-  if (!instName.empty()) {
-    loadInstrument<DataObjects::EventWorkspace>(instName, eventBuffer,
-                                                jsonGeometry);
-    if (rawMsgBuffer.empty()) {
-      eventBuffer->rebuildSpectraMapping();
+  auto instrument = eventBuffer->getInstrument();
+  if (instrument->getNumberDetectors() != eventBuffer->getNumberHistograms()) {
+    // Load the instrument if possible but continue if we can't
+    if (!instName.empty()) {
+      loadInstrument<DataObjects::EventWorkspace>(instName, eventBuffer,
```

Comment:
Would it be better to put the empty check into `loadInstrument` instead, or are there legitimate cases where we would want `loadInstrument` to throw.